File: app/src/main/java/com/forever/bee/akgraphql/ui/TripsFragment.java
package com.forever.bee.akgraphql.ui;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.apollographql.apollo.ApolloClient;
import com.apollographql.apollo.api.Response;
import com.apollographql.apollo.rx2.Rx2Apollo;
import com.forever.bee.akgraphql.BuildConfig;
import com.forever.bee.akgraphql.GetAllTrips;
import java.util.Objects;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import okhttp3.OkHttpClient;

/**
 * Build the [TripsFragment] class
 */
public class TripsFragment extends RecyclerViewFragment {
    //An instance of ApolloClient, this is a wrapper around an OkHttpClient that knows about Apollo-Android
    private final ApolloClient apolloClient = ApolloClient.builder()
            .okHttpClient(new OkHttpClient())
            .serverUrl(BuildConfig.GRAPHQL_API_KEY)
            .build();
    // An instance of RxJava Observable chains
    private Observable<GetAllTrips.Data> observable;
    private Disposable sub;


    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        // Setting up RxJava Observable chains using the Rx2Apollo helper class
        observable = Rx2Apollo.from(apolloClient.query(new GetAllTrips()).watcher())
                .subscribeOn(Schedulers.io())
                .map(this::getAllTripsFields)
                .cache()
                .observeOn(AndroidSchedulers.mainThread());

    }


    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        setLayoutManager(new LinearLayoutManager(getActivity()));

        getRecyclerView()
                .addItemDecoration(new DividerItemDecoration(getActivity(),
                        LinearLayoutManager.VERTICAL));

        unsub();

        sub = observable.subscribe(
                s -> setAdapter(buildRVAdapter(s)),
                error -> {
                    Toast.makeText(getActivity(), error.getMessage(), Toast.LENGTH_LONG).show();
                    Log.e(getClass().getSimpleName(), "Exception processing request",
                            error);
                }
        );
    }

    @Override
    public void onDestroy() {
        unsub();
        super.onDestroy();
    }

    private void unsub() {
        if (sub != null && !sub.isDisposed()) {
            sub.dispose();
        }
    }

    private GetAllTrips.Data getAllTripsFields(Response<GetAllTrips.Data> response) {
        if (response.hasErrors()) {
            throw new RuntimeException(Objects.requireNonNull(response.getErrors()).get(0).getMessage());
        }

        return response.getData();
    }

    private RecyclerView.Adapter buildRVAdapter(GetAllTrips.Data response) {
        return (new TripsAdapter(response.allTrips(), getActivity().getLayoutInflater(),
                android.text.format.DateFormat.getDateFormat(getActivity())));
    }
}

File: app/src/main/java/com/forever/bee/akgraphql/trips/TripsFragment.kt
package com.forever.bee.akgraphql.trips

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.Transformations.map
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import com.apollographql.apollo.ApolloClient
import com.apollographql.apollo.api.Response
import com.apollographql.apollo.rx2.Rx2Apollo
import com.forever.bee.akgraphql.BuildConfig
import com.forever.bee.akgraphql.GetAllTrips
import com.forever.bee.akgraphql.MainActivity
import com.forever.bee.akgraphql.databinding.FragmentTripsBinding
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import okhttp3.OkHttpClient

/**
 *
 * @property apolloClient  An instance of [ApolloClient]
 * */
class TripsFragment : Fragment() {
    private var _binding: FragmentTripsBinding? = null
    private val binding get() = _binding!!
    private lateinit var tripsAdapter: TripsAdapter
    private lateinit var callingActivity: MainActivity
    private lateinit var observable: Observable<GetAllTrips.Data>

    /*
    * This is a wrapper around an OkHttpClient that knows about Apollo-Android.
    * */
    private val apolloClient = ApolloClient.builder()
        .okHttpClient(OkHttpClient())
        .serverUrl(BuildConfig.GRAPHQL_API_KEY)
        .build()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        observable = Rx2Apollo.from(apolloClient.query(GetAllTrips()).watcher())
            .subscribeOn(Schedulers.io())
            .cache()
            .observeOn(AndroidSchedulers.mainThread())
            .map{response -> (getAllTripsFields(response))}
    }

    /**
     * Initialise the fragment_trips layout's binding class so the fragment can interact
     * with the layout's components.
     * */
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return super.onCreateView(inflater, container, savedInstanceState)
        callingActivity = activity as MainActivity
        _binding = FragmentTripsBinding.inflate(inflater, container, false)

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.recyclerView.layoutManager = LinearLayoutManager(activity)
        binding.recyclerView.itemAnimator = DefaultItemAnimator()

        // Initialise the TripsAdapter adapter
        tripsAdapter = TripsAdapter(callingActivity)
        binding.recyclerView.adapter = tripsAdapter
    }

    /**
     * Prevent the fragment from accessing components of the layout in the event the layout has been
     * closed but the fragment is not yet shut down.
     * */
    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    private fun getAllTripsFields(response: Response<GetAllTrips.Data>): GetAllTrips.Data? {
        if (response.hasErrors()) {
            throw RuntimeException(response.errors?.get(0)?.message ?: "Unknown Exception")
        }

        return response.data
    }
}


